Allows setting a height to `gr.File` and improves the UI of the component

abidlabs opened the pull request:
This PR:

* Adds a `height` parameter to `gr.File()` which sets both the height of upload box as well as the maximum height of the displayed files. If the number of files exceeds what can be displayed, then a scrollbar appears. This is useful when displaying a large number of files (and closes: #5131)
* Fixes some incorrect types in `gr.Gallery` which was the inspiration for this PR
* Improves the UI of `gr.File` a bit --> no more "Download" text. Now just clicking on the filename downloads the file. This looks better on smaller screens and also makes it easier for users to download the file that they intend to download since they no longer have to trace across the screen. 

<img width="770" alt="image" src="https://github.com/gradio-app/gradio/assets/1778297/7b237f81-96b0-46d7-a384-aa3176ca60cf">

* Adds stories for `gr.File`

As a side note, it seemed like I had to add the `height` prop to a lot of different places in order to get this to work. Maybe I'm doing something wrong? 

Fixes: #5131

pngwn commented:
Passing that prop down a few levels isn't really unexpected, there are alternative approaches we can take if it gets annoying though.We will be able to remove one layer when #5215 is merged.

dawoodkhan82 commented:
![Screenshot 2023-08-15 at 2 19 08 PM](https://github.com/gradio-app/gradio/assets/12725292/e6bbe12b-81c0-4add-a946-4ecaa6ff6b8b)

Looks like we need to add padding for the "x" button

dawoodkhan82 commented:
Also how do we feel about adding a drop shadow when scrolling is enabled to make it clear that the component can be scrolled.

aliabid94 commented:
Block.svelte takes a `height` attribute - see [Image](https://github.com/gradio-app/gradio/blob/0df9092e193ad6bdbaeff428adfa954d79aedb60/js/image/static/StaticImage.svelte#L41) component or [Chatbot](https://github.com/gradio-app/gradio/blob/0df9092e193ad6bdbaeff428adfa954d79aedb60/js/chatbot/static/StaticChatbot.svelte#L58-L59) component for examples. Shouldn't have to add it everywhere else.

abidlabs commented:
> Looks like we need to add padding for the "x" button. This is when I set height to 200px

Will fix, thanks

> Also how do we feel about adding a drop shadow when scrolling is enabled to make it clear that the component can be scrolled.

I can take a look at this -- do we have this in any other components?

abidlabs commented on js/file/shared/FilePreview.svelte:
```diff
@@ -41,12 +38,16 @@
 									: file.orig_name || file.name}
 							>
-								Download
+								{display_file_name(file)}
```
Thanks @hannahblair appreciate the suggestion. Putting the extension in all caps feels very visually crowded to me, particularly if there's a list of many files. What do you think about instead always including the extension at the end of the filename even if we trim? So it would look like:

* `important_document984023984903...pdf`

Shorter filenames would remain in their original form:

* `requirements.txt`

abidlabs replied:
Fixed:
<img width="630" alt="image" src="https://github.com/gradio-app/gradio/assets/1778297/cb8a5e6f-8fed-4866-8529-1a23ef9b2d83">

abidlabs commented:
@aliabid94 I couldn't get the Blocks height to work if the the `FilePreview` component was being used. Plus I think we'll need access to the `height` parameter inside the `FilePreview` component to address #5247 so I've left it as is for now.

@dawoodkhan82 I like the suggestion of adding a drop shadow effect -- this feedback applies to several components, so I've moved it to a separate issue: #5247.

Everything else should be addressed!

abidlabs commented:
This branch is starting to act super weird... will close and reopen a new PR 😅

C0untFloyd commented:
Thanks for implementing this, however can the removal of the download link be optional? Because I'm using the file component selection as filename input for an image component. Clicking on it with the fix would immediately start a download now, wouldn't it?

abidlabs commented:
> Thanks for implementing this, however can the removal of the download link be optional? Because I'm using the file component selection as filename input for an image component. Clicking on it with the fix would immediately start a download now, wouldn't it?

Very good point @C0untFloyd I'll redesign it bit

abidlabs commented:
Redesigned slightly based on the point that @C0untFloyd brought up.

<img width="1129" alt="image" src="https://github.com/gradio-app/gradio/assets/1778297/6e18c3cd-91c5-4c8c-8b8e-c02f5e348eb7">

Here's a demo to test both the select feature and downloading links:

```py
import gradio as gr

def s(i, x: gr.SelectData):
    return x.index

with gr.Blocks() as demo:
    gr.File()
    f = gr.File(["cheetah.jpg", "test.mp3"])
    t = gr.Textbox()
    f.select(s, f, t)
    
demo.launch()
```

Have merged in #5253, everything seems be working now!

dawoodkhan82 commented:
Nice, I like the new download button design! Also tested and every pretty much works great. 
Just a few UI nits:

(1) Setting the height to something really small (50px) kinda breaks the UI. Maybe we have a min-height?
<img width="1515" alt="Screenshot 2023-08-17 at 4 47 59 PM" src="https://github.com/gradio-app/gradio/assets/12725292/1ec8072b-567e-4ded-99f5-f9fea6701326">



(2) On a small screen size, the download icon gets pushed to a new line and looks awkward.
<img width="381" alt="Screenshot 2023-08-17 at 4 46 08 PM" src="https://github.com/gradio-app/gradio/assets/12725292/07285490-cf8c-4d95-a560-d3bc14a2036d">


(3) Hard to tell which download link is associated with a file, since there's so much white space in between. Not necessarily an issue with your changes, but maybe we open a new issue for this?
<img width="1520" alt="Screenshot 2023-08-17 at 4 46 53 PM" src="https://github.com/gradio-app/gradio/assets/12725292/88ab3843-1975-49fc-a13a-712126225b23">

abidlabs commented:
Thanks @dawoodkhan82 for the re-review!

(1) I think its fine to leave this to the whims of the developer. The same behavior is true when setting the height of the `Image` component to a very small value, for example, and it hasn't posed any issues.

(2) Ah yeah let me fix this

(3) Yeah I noticed this as well. Let me play around with something.

abidlabs commented:
Ok I fixed both of those issues @dawoodkhan82! For (3), I took a leaf from the `gr.Dataframe` and added an alternating background. See mockups below:

<img width="1304" alt="image" src="https://github.com/gradio-app/gradio/assets/1778297/3cd05da6-0f41-4cdf-8446-b35b618c4739">

Dark mode:

<img width="1318" alt="image" src="https://github.com/gradio-app/gradio/assets/1778297/66121da5-cd71-47f8-9c38-309634e5fa9e">

dawoodkhan82 commented:
LGTM very clean!

dawoodkhan82 commented:
> (1) I think its fine to leave this to the whims of the developer. The same behavior is true when setting the height of the Image component to a very small value, for example, and it hasn't posed any issues.

I feel like we should rethink this. We should always try and make it impossible (or at least very hard) for a dev to create a broken or ugly demo.

abidlabs commented:
> I feel like we should rethink this. We should always try and make it impossible (or at least very hard) for a dev to create a broken or ugly demo.

That's fair -- maybe we can tackle this as part of https://github.com/gradio-app/gradio/issues/5247, since both of these issues govern components whose `height` can be set

abidlabs commented:
Thanks for the detailed review @dawoodkhan82! Pretty happy with where we ended up

jkyndir commented:
i'm using gradio 4.11.0 in my project. but the height setting doesn't seem to take effect on gr.File at all?

any help would be appreciated

HuntZhaozq commented:
> i'm using gradio 4.11.0 in my project. but the height setting doesn't seem to take effect on gr.File at all?
> 
> any help would be appreciated

I meet the same problem. How to solve this ? @abidlabs

vernikouskaya commented:
Hi, I constantly update gradio, but since version 4.15 I used for the first time there is no scroller component appearing in gr.Files if the "height" parameter is set. This is my code line:
files_input = gr.Files(file_types=['image'], type='filepath', height=256, interactive=True)
I would really appreciate any help!

abidlabs commented:
Hi @vernikouskaya I can reproduce the problem. Would you be able to create a new issue for this?
